lib/bap_future/bap_future.ml

```diff
@@ -0,0 +1,497 @@
+      s'
+
+    let parse s ~init ~f =
+      let s',Signal publish = create () in
+      let state = ref init in
+      let step x =
+        let x,s = f !state x in
+        state := s;
+        Option.iter x ~f:publish in
+      link s s' step;
+      s'
+
+    let listen x f = ignore (subscribe x f)
+
+    let of_list xs = unfold_until ~init:xs ~f:(function
+        | [] -> None
+        | x::xs -> Some (x, xs))
+
+    let of_array xs =
+      let n = Array.length xs in
+      unfold_until ~init:0 ~f:(fun i ->
+          if i < n then Some (xs.(i), (i + 1))
+          else None)
+
+    let of_sequence xs = unfold_until ~init:xs ~f:Sequence.next
+
+    let zip = merge ~f:(fun x y -> x,y)
+
+    let repeat x = from (fun () -> x)
+
+    let sync ~clk dat =
+      let buf = Queue.create  () in
+      let stream, Signal push = create () in
+      link dat stream (fun x -> Queue.enqueue buf x);
+      link clk stream (fun () ->
+          let xs = Queue.to_list buf in
+          Queue.clear buf;
+          push xs);
+      stream
+
+    let apply fs xs = merge fs xs ~f:(fun f x -> f x)
+
+    let upon finished s : 'a future =
+      let result = ref None in
+      observe s (fun x -> result := Some x);
+      let future,promise = Future.create () in
+      Future.upon finished (fun () -> match !result with
+          | None -> ()
+          | Some x -> Promise.fulfill promise x);
+      future
+
+    let before event xs = 
+      let future, promise = Future.create () in
+      let buf = Queue.create () in     
+      let id = subscribe xs (Queue.enqueue buf) in
+      Future.upon event
+        (fun () -> 
+           unsubscribe xs id;
+           Promise.fulfill promise (Queue.to_list buf)); 
+      future
+
+    let last_before e xs n =
+      let buf = Queue.create () in
+      let add_value x = 
+        if Queue.length buf >= n then 
+          ignore (Queue.dequeue buf);
+        Queue.enqueue buf x in
+      let id = subscribe xs add_value in
+      let future, promise = Future.create () in
+      Future.upon e 
+        (fun () -> 
+           unsubscribe xs id;
+           Promise.fulfill promise (Queue.to_list buf));
+      future
+
+    let take xs n = 
+      let buf = Queue.create () in 
+      let future, promise = Future.create () in
+      let add_value id x = 
+        if Queue.length buf < n then Queue.enqueue buf x;
+        if Queue.length buf = n then
+          (unsubscribe xs id;
+           Promise.fulfill promise (Queue.to_list buf)) in
+      watch xs add_value;
+      future
+
+    let nth xs n = 
+      let cnt = ref 0 in
+      let future, promise = Future.create () in
+      let f id x = 
+        if !cnt = n then 
+          (unsubscribe xs id;
+           Promise.fulfill promise x);
+        Int.incr cnt; in
+      watch xs f;
+      future
+      
+    let find xs ~f = 
+      let future, promise = Future.create () in
+      let f id x = 
+        if f x then 
+          (unsubscribe xs id;
```

Comment:
the recurring pattern `unsubscribe and promise` is worthwhile to become a function. It occurs in at least 6 functions.